File: homework/Long_Nguyen/lesson_7/HouseRobberIII.java
import java.util.HashMap;
import java.util.Map;

// https://leetcode.com/problems/house-robber-iii
public class HouseRobberIII {
    public int rob(TreeNode root) {
        return solver(root, new HashMap<>());
    }

    private int solver(TreeNode root, Map<TreeNode, Integer> dp) {
        if (root == null)
            return 0;

        int valueIfRobCurrent = root.val;
        if (dp.containsKey(root))
            return dp.get(root);

        if (root.left != null) {
            valueIfRobCurrent += (solver(root.left.left, dp) + solver(root.left.right, dp));
        }
        if (root.right != null) {
            valueIfRobCurrent += (solver(root.right.left, dp) + solver(root.right.right, dp));
        }

        int valueIfNotRobCurrent = solver(root.left, dp) + solver(root.right, dp);
        int ans = Math.max(valueIfRobCurrent, valueIfNotRobCurrent);
        dp.put(root, ans);

        return ans;
    }
}

File: homework/Van_Le/homework7.1/HouseRobberIII.cs
public class HouseRobberIII
{
    public bool HasPathSum(TreeNode root, int targetSum)
    {
        public int Rob(TreeNode root)
        {
            var (withRoot, withoutRoot) = DFS(root);
            return Math.Max(withRoot, withoutRoot);
        }

        private (int, int) DFS(TreeNode root)
        {
            if (root == null) return (0, 0);

            var (leftWithRoot, leftWithoutRoot) = DFS(root.left);
            var (rightWithRoot, rightWithoutRoot) = DFS(root.right);

            var withRoot = root.val + leftWithoutRoot + rightWithoutRoot;
            var withoutRoot = Math.Max(leftWithRoot, leftWithoutRoot) + Math.Max(rightWithRoot, rightWithoutRoot);

            return (withRoot, withoutRoot);
        }
    }
}
